---
title: JS Promise
category: JS
tag: TIL
---

<iframe width="560" height="315" src="https://www.youtube.com/embed/3Ao3OroqQLQ" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
<iframe width="560" height="315" src="https://www.youtube.com/embed/Uh8u20MD978" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

## Promise

`resolve`는 성공, `reject`는 실패를 가리킨다.

```js
const pr = new Promise((resolve, reject) => {
  setTimeout(() => {
    resolve("OK")
  }, 1000);
});

console.log("start");
pr.then((result) => {
  console.log(result);
})
  .catch((err) => {
    console.log("err");
  })
  .finally(() => {
    console.log("end");
  });

// start
// OK
// end
```
promise가 resolve되면 `then`뒤의 구문을 수행, reject되면 `catch`를 통해 에러를 반환한다. `finally`는 성공여부에 관계없이 마지막에 무조건 수행된다.

## 콜백지옥

```js
const f1 = () => {
    setTimeout(function () {
        console.log("1번 주문 완료");
    }, 1000);
};

const f2 = (callback) => {
    setTimeout(function () {
        console.log("2번 주문 완료");
    }, 3000);
};

const f3 = (callback) => {
    setTimeout(function () {
        console.log("3번 주문 완료");
    }, 2000);
};

console.log("start")

//콜백 지옥

f1(function () {
    f2(function () {
        f3(function () {
            console.log("end")
        })
    })
})
```
이와 같이 `depth`가 깊어지면서 콜백함수를 통해 추가하는 것을 `콜백지옥`이라고 한다.

## 프로미스 체이닝

```js
const f1 = () => {
    return new Promise((res, rej) => {
        setTimeout(() => {
            res("1번 주문완료");
        }, 1000);
    });
};

const f2 = (message) => {
    console.log(message);
    return new Promise((res, rej) => {
        setTimeout(() => {
            res("2번 주문 완료");
        }, 3000);
    });
};

const f3 = (message) => {
    console.log(message);
    return new Promise((res, rej) => {
        setTimeout(() => {
            res("3번 주문 완료");
        }, 2000);
    });
};

console.log("start")

// 프로미스 체이닝

f1()
    .then((res) => f2(res))
    .then((res) => f3(res))
    .then((res) => console.log(res))
    .catch(console.log)
    .finally() => {
        console.log("end");
    };
```

## Promise.all

위의 코드에서 순차적으로 promise를 수행하면 총 6초가 걸린다. 이 시간을 더 짧게 줄이기위해 promise를 동시에 수행할 수 있게 하는 것이 `Promise.all`이다.

```js
Promise.all([f1(), f2(), f3()].then((res) => {
    console.log(res);
}));
```

동시에 실행하고자 하는 함수의 배열을 받아서 모두 완료되면 then 뒤의 코드가 처리된다. 순차적으로 수행했을 때 6초가 걸린 것을 동시에 실행해 시간을 3초로 단축시킬 수 있다.

## Promise.race

```js
Promise.race([f1(), f2(), f3().then((res) => {
    console.log(res)
})])
```

경주와 같아서 배열 안의 함수가 하나라도 완료되면 끝난다. 용량이 큰 이미지를 로딩할 때 하나라도 로딩이 되면 실행을 마치도록 사용할 수 있다. 제일 빨리 끝나는 (1초) f1함수가 실행을 마치면 전체 과정이 끝난다.

## async, await

```js
async function getName() {
    return "Juri";
}
```

await 옆엔 promise가 오고 이 promise가 완료될 때까지 기다린다.

```js

console.log("start")
async funtion order() {
    try{
    const result1 = await f1();
    const result2 = await f2(result1);
    const result3 = await f3(result2);
    console.log(result3);
    }catch(e){
        console.log(e)
    }
    console.log("end");
}

order();
```

프로미스 체이닝에서 `catch`로 error를 잡았다면, async-await 구문에서는 `try..catch`문을 사용해 error를 잡는다.

### promise.all

```js
async function order() {
    try{
        const result = await Promise.all([f1(), f2(), f3()])
        console.log(result);
    }catch(e) {
        console.log(e);
    }
}
```